tests/backend/riscv/test_register_allocation_constraints.py

```diff
@@ -0,0 +1,177 @@
+from abc import ABC
+from typing import Annotated, Generic, TypeAlias, TypeVar
+
+import pytest
+
+from xdsl.backend.riscv.register_allocation_constraints import OpTieConstraints
+from xdsl.dialects.riscv import (
+    AssemblyInstructionArg,
+    IntRegisterType,
+    Registers,
+    RISCVInstruction,
+    RISCVRegisterType,
+)
+from xdsl.ir import OpResult
+from xdsl.irdl import (
+    ConstraintVar,
+    Operand,
+    irdl_op_definition,
+    operand_def,
+    result_def,
+)
+from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.test_value import TestSSAValue
+
+RD1InvT = TypeVar("RD1InvT", bound=RISCVRegisterType)
+RD2InvT = TypeVar("RD2InvT", bound=RISCVRegisterType)
+RS1InvT = TypeVar("RS1InvT", bound=RISCVRegisterType)
+RS2InvT = TypeVar("RS2InvT", bound=RISCVRegisterType)
+
+
+class TestOp4(Generic[RD1InvT, RD2InvT, RS1InvT, RS2InvT], RISCVInstruction, ABC):
+
+    rd1: OpResult = result_def(RD1InvT)
+    rd2: OpResult = result_def(RD2InvT)
+    rs1: Operand = operand_def(RS1InvT)
+    rs2: Operand = operand_def(RS2InvT)
+
+    def assembly_line_args(self) -> tuple[AssemblyInstructionArg, ...]:
+        return self.rd1, self.rd2, self.rs1, self.rs2
```

Comment:
I was thinking you were going to do something a bit more like this:
```suggestion
class TestOp4(Generic[RdRsInvT, RD2InvT, RS2InvT], RISCVInstruction, ABC):

    SameIntRegisterType: TypeAlias = Annotated[RdRsInvT, ConstraintVar("RdRs")]

    rd1: OpResult = result_def(SameIntRegisterType)
    rd2: OpResult = result_def(RD2InvT)
    rs1: Operand = operand_def(SameIntRegisterType)
    rs2: Operand = operand_def(RS2InvT)

    def assembly_line_args(self) -> tuple[AssemblyInstructionArg, ...]:
        return self.rd1, self.rd2, self.rs1, self.rs2
```
So embedding in IRDL the fact that it's the same variable, as opposed to constraining the generic parameter itself.

Reply:
Much nicer! In this case tho I'm instantiating the generic class with different tie constraints for testing reasons, your suggestion would be 100% error proof when defining real world operations.